modules/python/src2/cv2.cpp

```diff
@@ -130,53 +130,162 @@ struct ConstDef
+        );
+    }
+
+    const std::string parent_name = PyModule_GetName(parent_module);
+
+    /// Special case handling when caller tries to register a submodule of the parent module with
+    /// the same name
+    if (name == parent_name) {
+        return parent_module;
+    }
+
+    if (!strStartsWith(name, parent_name))
     {
-        submod = PyImport_AddModule(full_name.c_str());
-        PyDict_SetItemString(d, short_name.c_str(), submod);
+        return PyErr_Format(PyExc_ImportError,
+            "Bindings generation error. "
+            "Submodule name should always start with a parent module name. "
+            "Parent name: %s. Submodule name: %s", parent_name.c_str(),
+            name.c_str()
+        );
     }
 
-    if (short_name != "")
-        root = submod;
-  }
-
-  // populate module's dict
-  PyObject * d = PyModule_GetDict(root);
-  for (PyMethodDef * m = methods; m->ml_name != NULL; ++m)
-  {
-    PyObject * method_obj = PyCFunction_NewEx(m, NULL, NULL);
-    PyDict_SetItemString(d, m->ml_name, method_obj);
-    Py_DECREF(method_obj);
-  }
-  for (ConstDef * c = consts; c->name != NULL; ++c)
-  {
-    PyDict_SetItemString(d, c->name, PyLong_FromLongLong(c->val));
-  }
+    size_t submodule_name_end = name.find('.', parent_name.size() + 1);
+    /// There is no intermediate submodules in the provided name
+    if (submodule_name_end == std::string::npos)
+    {
+        submodule_name_end = name.size();
+    }
 
+    PyObject* submodule = parent_module;
+
+    for (size_t submodule_name_start = parent_name.size() + 1;
+         submodule_name_start < name.size(); )
+    {
+        const std::string submodule_name = name.substr(submodule_name_start,
+                                                       submodule_name_end - submodule_name_start);
+
+        const std::string full_submodule_name = name.substr(0, submodule_name_end);
+
+
+        PyObject* parent_module_dict = PyModule_GetDict(submodule);
+        /// If submodule already exists it can be found in the parent module dictionary,
+        /// otherwise it should be added to it.
+        submodule = PyDict_GetItemString(parent_module_dict,
+                                         submodule_name.c_str());
+        if (!submodule)
+        {
+            submodule = PyImport_AddModule(full_submodule_name.c_str());
+            if (PyDict_SetItemString(parent_module_dict, submodule_name.c_str(), submodule) < 0) {
+                Py_CLEAR(submodule);
+                return PyErr_Format(PyExc_ImportError,
+                    "Can't register a submodule '%s' (full name: '%s')",
+                    submodule_name.c_str(), full_submodule_name.c_str()
+                );
+            }
+            /// PyDict_SetItemString doesn't steal a reference so the reference counter
+            /// of the submodule should be decremented to bind submodule lifetime to the
+            /// parent module
+            Py_DECREF(submodule);
```

Comment:
There are OSX logs before and after removal of this line:

- [python2_import.txt](https://github.com/opencv/opencv/files/7947920/python2_import.txt)
- [python2_import_2.txt](https://github.com/opencv/opencv/files/7947922/python2_import_2.txt)

Reply:
`PyImport_AddModule()` docs:
- Python3: https://docs.python.org/3/c-api/import.html
- Python2: https://docs.python.org/2/c-api/import.html

> Return value: Borrowed reference

and description:

> Return the module object corresponding to a module name.  The *name* argument
   may be of the form ``package.module``. First check **the modules dictionary** if
   there's one there, and if not, create a new one and insert it in the modules
   dictionary.

Likely "the modules dictionary" is a global `sys.modules`. So reference counter should be 2, not 1:
- one in sys.modules
- one after `PyDict_SetItemString`

Reply:
Oh, thank you. I see know the logic behind `PyImport_AddModule` and then `PyDict_SetItemString`. 

Should I fix the patch to 3.4 first, and then update this PR?

Reply:
These patches are independent (as we do manual port to 4.x for this patch). Update this and create separate PR to 3.4.